### PR TITLE
[GOVCMSD8-627] Prevent user roles from being configured as the admin role.

### PR DESCRIPTION
**Pull request by mingsong-hu**
In a SaaS site, no user role is allowed to be configured as an admin role.

This PR delivers following new features:

- Update all existing user roles to reset the 'is_admin' flag as false.
- Lock down the configuration of Administrator role from the UI (/admin/config/people/accounts)
- Prevent a user role from being configured as the admin role by decorating the core services of config.storage.

### CODE REVIEW

**Comment by simesy** [comment]
I think it makes sense to lock this down going forward. I've screenshot the setting to make it clear what this is about.

Screen Shot 2020-05-27 at 10 14 23

Unfortunately, applying this update will cause existing roles to lose access to things they actually need. Are you going to give these roles explicit access to all the permissions they might need? Or...

**Review by fubarhouse — APPROVED** [review]

**Comment by simesy** [comment]
So this is approved while the outcome will be that site administrators will be locked out of functionality?

**Comment by fubarhouse** [comment]
Users are still able to get the access when needed if it's in accordance with compliance regulations, but this will be delegated to the service desk. As for governance I'm not involved in that process... but it will be managed according to compliance requirements as has always been required - this change simply enforces rules that we're obligated to comply with.

* https://www.cyber.gov.au/ism
* https://www.protectivesecurity.gov.au/

Having said that I have seen the data around your concern - and there's far fewer people impacted than you might imagine (in regards to percentages)

Note that we are yet to put together comms around this change, so stay tuned for formal advice

**Comment by simesy** [comment]
> Note that we are yet to put together comms around this change, so stay tuned for formal advice

Great, that is needed I think and wouldn't it be good if this PR included a reference to a changlog change. I'm delivering solutions that are potentially effected by this and I would like to know what the rules are. You don't publish a list of allowed (or blacklisted) permissions at this stage so as vendors we just guess.

**Comment by simesy** [comment]
>Having said that I have seen the data around your concern

Why not just say "we will contact each project that is effected" if it's only a small number.

**Comment by brett-sargeant** [comment]
It should still be mentioned in a communication, maybe one of the agency updates.

**Review by fubarhouse — CHANGES_REQUESTED** [review]
This PR is to be resubmitted with changes to its scope.

**Comment by steveworley** [comment]
We just had a quick discussion around this proposed change. I think that the profile should not introduce the update hook to remove the `is_admin` flag from roles this would be a breaking change for many implementations and something that we should not promote as solutions for things. 

I would like to see feature flags and a configuration screen with sane defaults for the community so we can enforce certain things at a platform level, but give the freedom to change this behaviour.

The proposed solution for this change:

- Add a settings form for the govCM profile
- Add an option for GovCMS UI Tweaks (which will provide the form alter to disable the option)
- Use Drutiny to remediate permissions for SaaS projects to meet the permission matrix (I completely agree this needs to be circulated, if I'm honest - I don't even know what it is)
- Add additional guards in the GovCMS platform to ensure this matrix isn't changed

**Comment by simesy** [comment]
My +1 for centralising such checks in Drutiny.
And for the record some other thoughts:

-  [govcms-vet](https://github.com/govCMS/scaffold-tooling/blob/develop/scripts/govcms-vet) could do **fast** CI-time static checking of yml files.
- [govcms-audit](https://github.com/govCMS/scaffold-tooling/blob/develop/scripts/govcms-audit) can run the **slow** nightly drutiny audit at CI time or locally on demand.

**Comment by brett-sargeant** [comment]
I thought there was already a check on deploy for permissions as part of the CI process?

re the permissions list, I put in a ticket last year requesting a copy of what was white/blacklisted and for it to be added to the solutions list. It was added to the internal tracker for publishing, but as far as I can tell it hasn't been published yet. Some of the permissions were still being decided on.

One potential issue is keeping it up to date after image/module updates introduce/remove items.

**Comment by klonos** [comment]
If `is_admin` is already set (due to drutiny not checking for this before, or by omission or human error) on SaaS sites, then it should be revoked and locked down - it simply is not acceptable security-wise, as roles with that privilege can simply grant permissions like "administer permissions" or "administer modules" to themselves. This effectively breaks the whole SaaS paradigm.

I am saying this in full knowledge that there will be an increased number of tickets in the GovCMS Service Desk 😓  ...it simply is something that will need to be done. We should not be loosening security to accomodate for customer "discomfort". People that will lose their ability to perform certain tasks will need to follow this process:

- Raise a ticket, explaining which specific permissions they need (or which task they need to be able to perform)
- Service Desk will assess the requested permissions according to the whitelist/blacklist and either grant those permissions, or refuse them.
- If the customer believes that a permission that has been refused should be granted, then this is escalated to the Finance Security team for a final decision.
- If a new permission gets introduced (due to core/contrib updates), then it will need to be escalated to the Finance Security, so that it can be assessed and explicitly added to either the black or the white list.

When customers cannot self-serve when it comes to a certain task that needs to be done, we have the following in place:

- Raise a ticket, and ask App Support to perform the task (for example, we have previously been doing that for site name change)
- App Support can work with DevOps and Finance Security to use modules included in the distro, such as:
  - https://www.drupal.org/project/module_permissions (delegate managing modules that have been whitelisted)
  - https://www.drupal.org/project/config_perms (create custom permissions to allow further self-serving)

My 2c

**Comment by njwesterman** [comment]
@klonos from first-hand experience of the support levels provided by the service desk - I strongly disagree with this approach and your definition of SaaS! Anything along the lines of what you have suggested, the govCMS team needs to pay and migrate customers this impacts to a PaaS solution. This is changing the fundamentals and main selling points of what the SaaS solution was all about. It's as simple as that.

**Comment by brett-sargeant** [comment]
It is the admin role (e.g. the one that is used by uuid=1), not the site_administrator role that we're referring to isn't it?

If so then site-admin should be able to cover most things required in SaaS, with the exception of the exportable configuration items done locally and pushed.  

In which case it's a matter of making sure that site-admin (or other roles) haven't been given admin level permissions where they shouldn't.

**Comment by simesy** [comment]
> It is the admin role (e.g. the one that is used by uuid=1)

Nah it's any role. There is an option to supercharge any role so that it bypasses all permission checks. See the image in the first comment.

**Comment by klonos** [comment]
>  I strongly disagree with this approach and your definition of SaaS!

@westerbail it is not _MY_ approach/definition of SaaS. SaaS customers cannot enable modules, and cannot manage permissions - they only have access to the themes, and that's pretty much it. Their access to the admin UI is limited to site-building tasks (access to views/blocks/content types etc.), and content editing tasks. That's how it's always been; even on Acquia. Any task that has security implications is locked down; period.

This worked fine with GovCMS7 - it's with GovCMS 8 and the introduction of config management where things took a wrong turn ...meaning: loopholes allowing certain things to be done when they shouldn't be possible. Again, not my personal interpretation of things - that's how SaaS is meant to work by design from a security point of view. This view is not just from working in the GovCMS Service Desk, this is how things worked in the agency I work for as a dev, when it came to GovCMS SaaS customers.

>This is changing the fundamentals and main selling points of what the SaaS solution was all about.

Not sure what you are referring to here(?) ...do you have any online literature to point to that supports this, or is it your interpretation of things? I am more than happy to be proven wrong and pointed to sales material that suggests that SaaS customers are in fact allowed to enable/disable modules, and manage user permissions.

> govCMS team needs to pay and migrate customers this impacts to a PaaS solution.

Not my place to comment on that one, and not what was being discussed here. Please feel free to take that with the people you need to.

**Comment by njwesterman** [comment]
@klonos - yes can send on all the information you have requested offline, I think the majority of our project work is still on the Salsa confluence pages, but sadly I don't have access to that information anymore, but we do have copies. 

Either way, I think it is best to move on from this and allow this thread to continue on message. As a customer, I think it's perfectly reasonable to ensure we know exactly what abilities we are losing before the change is made. A lot of the work we do on SaaS is absolutely time-critical. If something we are doing now, moves to a support ticket, then we need to know of this in advance, so we can plan the budget, plan the releases, assess the impact etc..

**Comment by simesy** [comment]
> sales material that suggests that SaaS customers are in fact allowed to enable/disable modules

Is there any sales material that explains what is allowed and not allowed in SaaS?

**Comment by brett-sargeant** [comment]
We can enable/disable modules that are covered by the SaaS version of GovCMS as not everyone needs all of those modules.

We can't add/enable modules that are not on the list of supported modules. 

The list is in the solutions section of the support site, though it hasn't been updated since September 2019, so it will be missing some of the modules that have been added since then.

What we need is for a list of what permissions can and can't be set to also be made available there.

**Comment by njwesterman** [comment]
@simesy I've requested the SaaS / PaaS decision matrix we went through with the material Salsa provided that led us down the SaaS route. I know it said we can enable/disable any module in the SaaS distro, however, you would need to go down the PaaS route if you wanted to add modules that aren't in the SaaS base. Additionally, we have the workflow security matrix built for SaaS that we administer, we have the ability to say only certain users can add a 'blog' and lower-level things like add but not remove certain media types etc.. I'm not too sure if my understanding of user permissions in this context for SaaS is right though, but the ability to set up and drive all of that was definitely part of it. 

I'll drop out of this one now, our biggest concerns were around what it is we actually lose, what then moves to a request to the service desk, what the service desks SLAs will be to review and action these new requests etc... so I'm just coming at it from what discomfort we may experience in our project management and planning activities as suppose to the technical details. I will leave you guys to it! Cheers!

**Comment by brett-sargeant** [comment]
@westerbail There was a document added to the solutions list under the SaaS customers section on 10 June that outlines the permissions.

**Comment by simesy** [comment]
Link as follows, just log into freshdesk to see. https://www.govcms.support/support/solutions/articles/51000046790-govcms-saas-permission-restrictions

**Comment by njwesterman** [comment]
Thanks, Guys - This does to my mind completely misrepresent the whole govCMS 2.0, the power is in your hand's ideology we were sold on, and the frustration comes from knowing that if we had this documentation less than a year ago, we would have gone PaaS all day long. 

I'm still trying to get my head around how it is going to affect us internally. 

Can anyone let me know what would change once these SaaS 'loopholes' are closed in this scenario...
 
We have an internal department developer working on the below project...

* Developer cuts a copy of the develop branch onto their machine for local development *

1. Adds in a taxonomy type
2. Adds in a new content type, however only certain users can create this content type and it needs to be reviewed by the head of that team - two new roles are created for author and approver, and the content moderation permissions are set for those roles
3. The RESTFul web services module is enabled
4. A JSON new view is built 
5. A ReactJS bundle is added to the custom theme that will be used to display the view using the data consumed from the JSON view endpoint
6. A custom template is added to the theme that the react build attaches too. 

The developer does an ahoy cex - checks in the config and theme - pushing to dev. It goes through test and is released to production where the dev then makes sure the correct people have the new roles etc..

Now we do all of this internally today - how do we do this in the upcoming new world - what changes?

**Comment by brett-sargeant** [comment]
From what I can see nothing changes, it's just the admin permissions on the remote site that is being changed, not local. It removes a security vulnerability from the live site. You can still do what is needed in the local development environment and export it.

**Comment by simesy** [comment]
I agree nothing changes. My rough checklist:

1. Make sure that no role has `is_admin: true` in config.
2. Make sure no role has `administer modules`, `administer permissions` and `administer software updates` (these typed from memory)

You can still change other permissions and enable modules via config deployment (for example, if enabled).

**Comment by njwesterman** [comment]
@simesy @brett-sargeant  - We had those is_admin changes ready to go in our test environment, but Salsa have just gone straight into master and made the changes themselves. 

So just in case the info helps on prod site - the extend menu is a 404 now, permissions menu is a 404 - People...you can still add by the looks of it - but you can't assign roles or change existing person's roles, so assuming that is now a govcms ticket.

**Comment by simesy** [comment]
@westerbail correct, unless you're using config deployments. https://www.lilengine.co/articles/govcms-new-ci-checks-roles-and-permissions

**Comment by simesy** [comment]
Yeap, you need to ask govcms help desk to change a user's roles.

**Comment by brett-sargeant** [comment]
If you can see the people screen then you can assign a role to a user https://sitenameaddress/admin/people

There should be checkboxes on the left hand side 
image

You select the name, then choose the role from the dropdown list 
image

Then apply to selected user.

**Comment by brett-sargeant** [comment]
There is one other thing that I can't recall if it was mentioned (and haven't time to reread the thread) config deployments will need to be enabled on your repository. 

Apparently it was causing issues for a number of agencies with configs being overwritten, so it's now off by default unless you ask for it.

**Comment by njwesterman** [comment]
Cool gotcha! Cheers guys!

**Pull request closed without merging**
